server/src/main/java/io/spine/server/enrich/EnricherBuilder.java

```diff
@@ -20,89 +20,178 @@
 
 package io.spine.server.enrich;
 
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.ImmutableSet;
-import com.google.common.collect.Sets;
+import com.google.errorprone.annotations.Immutable;
+import com.google.protobuf.Message;
 import io.spine.core.EnrichableMessageContext;
 
-import java.util.Optional;
-import java.util.Set;
-import java.util.function.BiFunction;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
 
+import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
-import static io.spine.util.Exceptions.newIllegalArgumentException;
+import static io.spine.type.MessageClass.interfacesOf;
 
 /**
- * The {@code Builder} allows to register {@link EnrichmentFunction}s handled by
- * the {@code Enricher} and set a custom translation function, if needed.
+ * Allows to register enrichment functions used by the {@link Enricher}.
  */
-public final class EnricherBuilder {
+public abstract class EnricherBuilder<M extends Message,
+                                      C extends EnrichableMessageContext,
+                                      B extends EnricherBuilder<M, C, B>> {
 
-    /** Functions which perform the enrichment. */
-    private final Set<EnrichmentFunction<?, ?, ?>> functions = Sets.newHashSet();
+    private static final String SUGGEST_REMOVAL = " Please call `remove(Class, Class)` first.";
+
+    /**
+     * Maps a pair of [source class, enrichment class] to a function which produces this enrichment.
```

Comment:
Isn't `{key : value}` the default way to express pairs? AFAIK `[value, value]` is generally understood as a tuple.